### ISSUE TITLE
Add apache pulsar based indexing service

### ISSUE THREAD

**Issue by aahmed-se** [apache/druid#7030]
Be able to index records by reading a set of pulsar topics.

https://pulsar.apache.org/

**Comment by gianm** [comment]
I am not too familiar with Pulsar, but if it has a partition / offset based scheme like Kafka or Kinesis, then it should be pretty straightforward to add using the same framework.

**Comment by aahmed-se** [comment]
it has a similar concept with cursors.

https://streaml.io/blog/cursors-in-pulsar

**Comment by niketh** [comment]
We have use cases to integrate with our internal version of Pulsar. @aahmed-se Are you working on this? If not, I was going to take a stab at it.

**Comment by aahmed-se** [comment]
@niketh it's on my radar, but don't have the bandwidth right now. You can go ahead.

**Comment by niketh** [comment]
It makes sense to implement this as a separate task and supervisor service. The characteristics of pulsar are different as compared to kafka. For example running pulsar in a shared subscription, the consuers can read messages from any parition. The need for maintaining per consumer offsets etc are not needed.

**Comment by niketh** [comment]
@aahmed-se Sounds good, I will take a stab at it.

**Comment by chariot1498** [comment]
Hey all my use case also needed to integrate druid with pulsar . Wanted to know did you guys have any luck doing that ?

**Comment by joshuadunham** [comment]
Anything I can do to help out with this? I'm also very interested. :)

**Comment by aahmed-se** [comment]
@joshuadunham You can take a stab and implementing it bet to look at kinesis or kafka ingest integration in druid and replicate it , it should not be that difficult.

**Comment by rueian** [comment]
Hi, I am working on this and having some progress with the Reader interface of pulsar client.

https://github.com/rueian/incubator-druid/commits/pulsar-indexing-service

May I create a PR after finishing the test cases?

**Comment by aahmed-se** [comment]
@rueian are you making changes to pulsar for this , if not then the code should be posted in druid itself.

**Comment by gianm** [comment]
@rueian Please do go ahead and open a PR, especially if you have something working & well tested. Thanks for your interest in contributing!

**Comment by haris-zynka** [comment]
Any updates on this one? I'm new to both Druid (recently heard about it never used it) and Apache Pulsar (this one at least I used), and I would love to do a test run but I heavily depend on Pulsar. If this is forgotten about I might research myself and try to do something but it's gonna take a lot of time until I figure things.

**Comment by rueian** [comment]
Sorry for no follow up from me.
However I am still blocked by something else. Please feel free to step in.

**Comment by devinbost** [comment]
@niketh Were you able to make any progress on this? My team is very interested in this capability.

**Comment by josephglanville** [comment]
I'm interested in taking this up.

My research leads to believe the best way to implement this is building on the SeekableStream abstraction. Now on the surface this may appear like an impedance mismatch as Pulsar is primarily built around managing offsets/consumer state on the broker side but I still think the SeekableStream approach is best for Druid because it best suits it's notions of tasks and segments.

The way I think this should be implemented is to have the supervisor create a task per partition of the Pulsar topic, each task will then use an exclusive, non-durable subscription that consumes from that specific partition. In this way seeking to a specific message ID can be supported cleanly, which is required to support task resumption and idempotency.

This will result in one segment per task however so users of this indexing service will likely want to enable compaction.

@sijie does this approach sound correct to you?

/cc @gianm @jsun98 @dclim as you gentlemen worked on the SeekableStream abstraction and would appreciate your thoughts.

**Comment by sijie** [comment]
@josephglanville Yes. The approach looks right to me.

An alternative approach is to integrate Druid with Pulsar via [Kafka-on-Pulsar](https://github.com/streamnative/kop) by leverage Druid's existing Kafka integration.

**Comment by devinbost** [comment]
> each task will then use an exclusive, non-durable subscription that consumes from that specific partition

Does this mean the topic would be non-persistent? I'm wanting to better understand what is meant by "non-durable" subscription.

**Comment by dat-vikash** [comment]
@devinbost The topic pesistence is set at creation 

Persistent Topic: `persistent://property/cluster/namespace/topic`  
  All messages will be stored on disk

Non persistent: `non-persistent://property/cluster/namespace/topic`  
    Messages will not be persisted to Disk

The durability refers to the cursor persistence. `Durable` subscriptions have the cursor persisted 
>If a broker restarts from a failure, it can recover the cursor from the persistent storage (BookKeeper), so that messages can continue to be consumed from the last consumed position.

While `non-durable` the cursor is lost on bookie restart. 
>Once a broker stops, the cursor is lost and can never be recovered, so that messages can not continue to be consumed from the last consumed position.

Since druid's supervisor creates short lived tasks and we want to support resumption from any point in the stream, we require a `non-durable` subscription mode. `durable` subscription cursor position is enforced by the bookie and cannot be overridden, meaning we could not support resumption from any point in the stream.  So to properly support resiliency and idempotent resumption, we need to use the `non-durable` subscription.

**Comment by devinbost** [comment]
We'd still like this feature.
